### DISCUSSION TITLE
How to pass arguments to class provider annotated with Riverpod annotation

### DISCUSSION THREAD

**Discussion by vorant94** [rrousselGit/riverpod#3102] [Q&A] [answered]
Hi, there

In order to perform side-effect on the app state the documentation uses annotated class example
```dart
@riverpod
class TodoList extends _$TodoList {
  @override
  Future<List<Todo>> build() async => [/* ... */];

  Future<void> addTodo(Todo todo) async {
    await http.post(
      Uri.https('your_api.com', '/todos'),
      // We serialize our Todo object and POST it to the server.
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(todo.toJson()),
    );
  }
}
```

In order to pass parameters to the provider the documentation uses annotated function example
```dart
@riverpod
Future<Activity> activity(
  ActivityRef ref,
  // We can add arguments to the provider.
  // The type of the parameter can be whatever you wish.
  String activityType,
) async {
  // We can use the "activityType" argument to build the URL.
  // This will point to "https://boredapi.com/api/activity?type=<activityType>"
  final response = await http.get(
    Uri(
      scheme: 'https',
      host: 'boredapi.com',
      path: '/api/activity',
      // No need to manually encode the query parameters, the "Uri" class does it for us.
      queryParameters: {'type': activityType},
    ),
  );
  final json = jsonDecode(response.body) as Map<String, dynamic>;
  return Activity.fromJson(json);
}
```

I wonder how can I combine both performing side-effect and passing arguments. Like for example I'd like to create some `UserDetailState`, that provides support for fetching and updating the user. 

One solution could be to return `null` in `build` method initially, then both fetch and update stuff should be performed as a side-effect methods. But this way `null` becomes a valid state case, which is not true in the reality... And if I should use annotated function syntax, I didn't find in the docs how can I perform side-effect on it like updating the user? 

so what is the recommended way to deal with it?

**Comment by rrousselGit** [comment, marked as answer]
Parameters should be added on the `build` method:

```dart
@riverpod
class TodoList extends _$TodoList {
  @override
  Future<List<Todo>> build(String activityType) {...}

}
```
